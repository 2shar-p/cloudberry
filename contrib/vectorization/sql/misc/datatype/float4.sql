
-- set vectorization
SET vector.enable_vectorization=off;

-- create table

CREATE TABLE FLOAT4_TBL (f1  float4) WITH(APPENDONLY=true, ORIENTATION=column);

INSERT INTO FLOAT4_TBL(f1) VALUES ('    0.0');
INSERT INTO FLOAT4_TBL(f1) VALUES ('1004.30   ');
INSERT INTO FLOAT4_TBL(f1) VALUES ('     -34.84    ');
INSERT INTO FLOAT4_TBL(f1) VALUES ('1.2345678901234e+20');
INSERT INTO FLOAT4_TBL(f1) VALUES ('1.2345678901234e-20');

-- test for over and under flow
INSERT INTO FLOAT4_TBL(f1) VALUES ('10e70');
INSERT INTO FLOAT4_TBL(f1) VALUES ('-10e70');
INSERT INTO FLOAT4_TBL(f1) VALUES ('10e-70');
INSERT INTO FLOAT4_TBL(f1) VALUES ('-10e-70');

-- bad input
INSERT INTO FLOAT4_TBL(f1) VALUES ('');
INSERT INTO FLOAT4_TBL(f1) VALUES ('       ');
INSERT INTO FLOAT4_TBL(f1) VALUES ('xyz');
INSERT INTO FLOAT4_TBL(f1) VALUES ('5.0.0');
INSERT INTO FLOAT4_TBL(f1) VALUES ('5 . 0');
INSERT INTO FLOAT4_TBL(f1) VALUES ('5.   0');
INSERT INTO FLOAT4_TBL(f1) VALUES ('     - 3.0');
INSERT INTO FLOAT4_TBL(f1) VALUES ('123            5');

-- set vectorization
SET vector.enable_vectorization=on;

-- queries
SELECT '' AS five, * FROM FLOAT4_TBL;

SELECT '' AS four, f.* FROM FLOAT4_TBL f WHERE f.f1 <> '1004.3';

SELECT '' AS one, f.* FROM FLOAT4_TBL f WHERE f.f1 = '1004.3';

SELECT '' AS three, f.* FROM FLOAT4_TBL f WHERE '1004.3' > f.f1;

SELECT '' AS three, f.* FROM FLOAT4_TBL f WHERE  f.f1 < '1004.3';

SELECT '' AS four, f.* FROM FLOAT4_TBL f WHERE '1004.3' >= f.f1;

SELECT '' AS four, f.* FROM FLOAT4_TBL f WHERE  f.f1 <= '1004.3';

SELECT '' AS three, f.f1, f.f1 * '-10' AS x FROM FLOAT4_TBL f
   WHERE f.f1 > '0.0';

SELECT '' AS three, f.f1, f.f1 + '-10' AS x FROM FLOAT4_TBL f
   WHERE f.f1 > '0.0';

SELECT '' AS three, f.f1, f.f1 / '-10' AS x FROM FLOAT4_TBL f
   WHERE f.f1 > '0.0';

SELECT '' AS three, f.f1, f.f1 - '-10' AS x FROM FLOAT4_TBL f
   WHERE f.f1 > '0.0';

-- test divide by zero
SELECT '' AS bad, f.f1 / '0.0' from FLOAT4_TBL f;

SELECT '' AS five, * FROM FLOAT4_TBL;

-- test the unary float4abs operator
SELECT '' AS five, f.f1, @f.f1 AS abs_f1 FROM FLOAT4_TBL f;

UPDATE FLOAT4_TBL
   SET f1 = FLOAT4_TBL.f1 * '-1'
   WHERE FLOAT4_TBL.f1 > '0.0';

SELECT '' AS five, * FROM FLOAT4_TBL;

-- drop table
DROP TABLE FLOAT4_TBL;
