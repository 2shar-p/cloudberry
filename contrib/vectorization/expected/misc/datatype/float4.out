
-- set vectorization
SET vector.enable_vectorization=off;
SET

-- create table

CREATE TABLE FLOAT4_TBL (f1  float4) WITH(APPENDONLY=true, ORIENTATION=column);
CREATE

INSERT INTO FLOAT4_TBL(f1) VALUES ('    0.0');
INSERT 1
INSERT INTO FLOAT4_TBL(f1) VALUES ('1004.30   ');
INSERT 1
INSERT INTO FLOAT4_TBL(f1) VALUES ('     -34.84    ');
INSERT 1
INSERT INTO FLOAT4_TBL(f1) VALUES ('1.2345678901234e+20');
INSERT 1
INSERT INTO FLOAT4_TBL(f1) VALUES ('1.2345678901234e-20');
INSERT 1

-- test for over and under flow
INSERT INTO FLOAT4_TBL(f1) VALUES ('10e70');
ERROR:  value out of range: overflow
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('10e70');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('-10e70');
ERROR:  value out of range: overflow
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('-10e70');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('10e-70');
ERROR:  value out of range: underflow
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('10e-70');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('-10e-70');
ERROR:  value out of range: underflow
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('-10e-70');
                                           ^

-- bad input
INSERT INTO FLOAT4_TBL(f1) VALUES ('');
ERROR:  invalid input syntax for type real: ""
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('       ');
ERROR:  invalid input syntax for type real: "       "
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('       ');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('xyz');
ERROR:  invalid input syntax for type real: "xyz"
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('xyz');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('5.0.0');
ERROR:  invalid input syntax for type real: "5.0.0"
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('5.0.0');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('5 . 0');
ERROR:  invalid input syntax for type real: "5 . 0"
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('5 . 0');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('5.   0');
ERROR:  invalid input syntax for type real: "5.   0"
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('5.   0');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('     - 3.0');
ERROR:  invalid input syntax for type real: "     - 3.0"
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('     - 3.0');
                                           ^
INSERT INTO FLOAT4_TBL(f1) VALUES ('123            5');
ERROR:  invalid input syntax for type real: "123            5"
LINE 1: INSERT INTO FLOAT4_TBL(f1) VALUES ('123            5');
                                           ^

-- set vectorization
SET vector.enable_vectorization=on;
SET

-- queries
SELECT '' AS five, * FROM FLOAT4_TBL;
 five | f1
------+-------------
      | -34.84
      | 1.23457e-20
      | 0.0
      | 1004.3
      | 1.23457e+20
(5 rows)

SELECT '' AS four, f.* FROM FLOAT4_TBL f WHERE f.f1 <> '1004.3';
 four | f1
------+-------------
      | 0.0
      | 1.23457e+20
      | -34.84
      | 1.23457e-20
(4 rows)

SELECT '' AS one, f.* FROM FLOAT4_TBL f WHERE f.f1 = '1004.3';
 one | f1
-----+--------
     | 1004.3
(1 row)

SELECT '' AS three, f.* FROM FLOAT4_TBL f WHERE '1004.3' > f.f1;
 three | f1
-------+-------------
       | -34.84
       | 1.23457e-20
       | 0.0
(3 rows)

SELECT '' AS three, f.* FROM FLOAT4_TBL f WHERE  f.f1 < '1004.3';
 three | f1
-------+-------------
       | -34.84
       | 1.23457e-20
       | 0.0
(3 rows)

SELECT '' AS four, f.* FROM FLOAT4_TBL f WHERE '1004.3' >= f.f1;
 four | f1
------+-------------
      | 0.0
      | 1004.3
      | -34.84
      | 1.23457e-20
(4 rows)

SELECT '' AS four, f.* FROM FLOAT4_TBL f WHERE  f.f1 <= '1004.3';
 four | f1
------+-------------
      | 0.0
      | 1004.3
      | -34.84
      | 1.23457e-20
(4 rows)

SELECT '' AS three, f.f1, f.f1 * '-10' AS x FROM FLOAT4_TBL f WHERE f.f1 > '0.0';
 three | f1          | x
-------+-------------+--------------
       | 1004.3      | -10043.0
       | 1.23457e+20 | -1.23457e+21
       | 1.23457e-20 | -1.23457e-19
(3 rows)

SELECT '' AS three, f.f1, f.f1 + '-10' AS x FROM FLOAT4_TBL f WHERE f.f1 > '0.0';
 three | f1          | x
-------+-------------+-------------
       | 1004.3      | 994.3
       | 1.23457e+20 | 1.23457e+20
       | 1.23457e-20 | -10.0
(3 rows)

SELECT '' AS three, f.f1, f.f1 / '-10' AS x FROM FLOAT4_TBL f WHERE f.f1 > '0.0';
 three | f1          | x
-------+-------------+--------------
       | 1.23457e-20 | -1.23457e-21
       | 1004.3      | -100.43
       | 1.23457e+20 | -1.23457e+19
(3 rows)

SELECT '' AS three, f.f1, f.f1 - '-10' AS x FROM FLOAT4_TBL f WHERE f.f1 > '0.0';
 three | f1          | x
-------+-------------+-------------
       | 1004.3      | 1014.3
       | 1.23457e+20 | 1.23457e+20
       | 1.23457e-20 | 10.0
(3 rows)

-- test divide by zero
SELECT '' AS bad, f.f1 / '0.0' from FLOAT4_TBL f;
ERROR:  division by zero

SELECT '' AS five, * FROM FLOAT4_TBL;
 five | f1
------+-------------
      | -34.84
      | 1.23457e-20
      | 0.0
      | 1004.3
      | 1.23457e+20
(5 rows)

-- test the unary float4abs operator
SELECT '' AS five, f.f1, @f.f1 AS abs_f1 FROM FLOAT4_TBL f;
 five | f1          | abs_f1
------+-------------+-------------
      | 0.0         | 0.0
      | 1004.3      | 1004.3
      | 1.23457e+20 | 1.23457e+20
      | -34.84      | 34.84
      | 1.23457e-20 | 1.23457e-20
(5 rows)

UPDATE FLOAT4_TBL SET f1 = FLOAT4_TBL.f1 * '-1' WHERE FLOAT4_TBL.f1 > '0.0';
UPDATE 3

SELECT '' AS five, * FROM FLOAT4_TBL;
 five | f1
------+--------------
      | 0.0
      | -1.23457e+20
      | -1.23457e-20
      | -34.84
      | -1004.3
(5 rows)

-- drop table
DROP TABLE FLOAT4_TBL;
DROP
